### ISSUE TITLE
improve condition-functions

### ISSUE THREAD

**Issue by L3MON4D3** [L3MON4D3/LuaSnip#936]
It would be cool to pass the snippet to the condition-functions, for example for checking whether the trigger matches in `show_condition` (inspired by #931) or doing other computations which require the snippet.
We can't just pass the snippet after the other arguments we currently have because we explicitly assure that a `show_condition` can be passed to the `condition`-option, but this wouldn't work if done this way (signature of `condition` would be `(line_to_cursor, matched_trigger, captures, snip)`, and `show_condition` would be `(line_to_cursor, snip)`, so the `show_condition` would receive `matched_trigger` in place of `snip` when used as a `condition`).

IMO the best solution for this is to add new options, `cond` and `showCond` (this way they will also finally matche the camelCase of other options :D) which just receive one table containing the available data, so `snip`, `line_to_cursor` for `showCond`, and additionally `matched_trigger` and `captures` for `cond` (this will be easy to extend!).
I'd like to add this in a non-breaking way, so we should (at least for a while, probably indefinitely) still accept `(show_)condition`, which will only require a bit of plumbing, but also make it so that code using `show_condition` (like [cmp_luasnip](https://github.com/saadparwaiz1/cmp_luasnip/blob/master/lua/cmp_luasnip/init.lua#L110)) will continue to work (it obviously won't work with a `showCond` requiring `snip`, but if that is communicated clearly all should be fine).

(just some more thoughts:) In order to ensure a smooth transition we should expose some kind of feature-list, where clients can check if, for example, the available luasnip-version has the new `showCond` (preferable immediately in `require("luasnip")`, so there doesn't have to be an additional check for the feature-list :D).

All in all, this is not very pressing, but I'd like to track this in an issue.
If there are usecases where this will come in handy, comment below, maybe it'll speed this up :P

**Comment by connorwyatt** [comment]
This is something I have come across recently when trying to define treesitter_postfix snippets. Because the snippet has a query that looks something like this:

```
[
    (identifier)
    (expression_statement)
] @prefix
```

With the current parameters to show_condition, I can't easily check that the treesitter query has been satisfied, which means that I get a lot of postfix snippets showing in my blink completion that aren't actually applicable

It's also possible that I'm approaching the problem incorrectly

**Comment by L3MON4D3** [comment]
Good point, it's definitely feasible to extend `treesitter-postfix` to include a `show_condition`, which does this check 🤔
